Copy of Participation Activity
# Coding exercises
Exercises 1-3 are thought exercises that don't require coding.

1. Explore the Jupyter Lab interface and look at some of the shortcuts available. Don't worry about memorizing them now (eventually they will become second nature and save you a lot of time), just get comfortable using notebooks.

2. Are all data normally distributed? Not all data is normally distrubuted. Data can be uneven and skewed with outliers.

3. When would it make more sense to use the median instead of the mean for the measure of center? When there are extreme outliers in data. For example if we're attempting to find the average height of a group of 100 people, but there is one person over 7 feet tall.

## Exercise 4: Generate the data by running this cell
This will give you a list of numbers to work with in the remaining exercises.

In [ ]:
import random
random.seed(0)
salaries = [round(random.random()*1000000, -3) for _ in range(100)]
print(list(salaries))

[844000.0, 758000.0, 421000.0, 259000.0, 511000.0, 405000.0, 784000.0, 303000.0, 477000.0, 583000.0, 908000.0, 505000.0, 282000.0, 756000.0, 618000.0, 251000.0, 910000.0, 983000.0, 810000.0, 902000.0, 310000.0, 730000.0, 899000.0, 684000.0, 472000.0, 101000.0, 434000.0, 611000.0, 913000.0, 967000.0, 477000.0, 865000.0, 260000.0, 805000.0, 549000.0, 14000.0, 720000.0, 399000.0, 825000.0, 668000.0, 1000.0, 494000.0, 868000.0, 244000.0, 325000.0, 870000.0, 191000.0, 568000.0, 239000.0, 968000.0, 803000.0, 448000.0, 80000.0, 320000.0, 508000.0, 933000.0, 109000.0, 551000.0, 707000.0, 547000.0, 814000.0, 540000.0, 964000.0, 603000.0, 588000.0, 445000.0, 596000.0, 385000.0, 576000.0, 290000.0, 189000.0, 187000.0, 613000.0, 657000.0, 477000.0, 90000.0, 758000.0, 877000.0, 923000.0, 842000.0, 898000.0, 923000.0, 541000.0, 391000.0, 705000.0, 276000.0, 812000.0, 849000.0, 895000.0, 590000.0, 950000.0, 580000.0, 451000.0, 660000.0, 996000.0, 917000.0, 793000.0, 82000.0, 613000.0, 486000.0]


## Exercise 5: Calculating statistics and verifying

Use the data generated above to calulate in code the following statistics without importing the statistics module. Then use the statistics module to verify your results.  Import the statistics module from https://docs.python.org/3/library/statistics.html

### mean

In [ ]:
mean_manual = sum(salaries) / len(salaries)
print(mean_manual)

585690.0


In [ ]:
import statistics
mean_verified = statistics.mean(salaries)
print(mean_manual)

585690.0


This means the average salary is $585,690

### median

In [ ]:
sorted_salaries = sorted(salaries)
n = len(sorted_salaries)

median_manual = (sorted_salaries[n // 2-1] + sorted_salaries[n // 2]) / 2

print(median_manual)

589000.0


In [ ]:

median_verified = statistics.median(salaries)
print(median_verified)

589000.0


This means the middle salary is 589,000. Half of the salaries are above and the other half are below this $589,000.

### mode

In [ ]:
mode_manual = max(set(salaries), key=salaries.count)
print(mode_manual)

477000.0


In [ ]:
mode_verified = statistics.mode(salaries)
print(mode_verified)

477000.0


The most frequently occuring salary is 477,000

### sample variance
Remember to use Bessel's correction.

Note: Bessel's Correction is n-1

In [ ]:
n = len(salaries)

variance_manual = sum((salary-mean_manual) ** 2 for salary in salaries) / (n-1)

print(variance_manual)

70664054444.44444


In [ ]:
variance_verified = statistics.variance(salaries)
print(variance_verified)

70664054444.44444


This is a measure of how spread out the salaries are around the mean. Not really sure how beneficial this is here. But this number is expressed in squared dollars and helps us with the next step.

### sample standard deviation
Remember to use Bessel's correction.

In [ ]:
standard_deviation_manual = variance_manual ** 0.5
print(standard_deviation_manual)

265827.11382484


In [ ]:
standard_deviation_verified = statistics.stdev(salaries)
print(standard_deviation_verified)

265827.11382484


This mean the usual distance between the salaries and the mean  of $585,690 is approximately $265,827.

## Exercise 6: Calculating more statistics
### range

In [ ]:
minimum_salary = min(salaries)
maximum_salary = max(salaries)
salary_range = maximum_salary - minimum_salary

print("Minimum:", minimum_salary)
print("Maximum:", maximum_salary)
print("Range:", salary_range)

Minimum: 1000.0
Maximum: 996000.0
Range: 995000.0


Range gives a picture of the spread, but it only uses the most extreme values. Meaning it can be heavily affected by outliers.

### interquartile range

In [ ]:
q1 = (sorted_salaries[24] + sorted_salaries[25]) / 2
q3 = (sorted_salaries[74] + sorted_salaries[75]) / 2

interquartile_range = q3 - q1

print("Q1:" , q1)
print("Q3:" , q3)
print("IQR:", interquartile_range)

Q1: 402000.0
Q3: 819500.0
IQR: 417500.0


IQR measures the spread of the middle 50% of salaries: IQR = q3 - q1. q1 is center of the lower half, while q3 is the center of the upper half.
Results Meaning:
About 25% of salaries are at or below 402,000
About 25% of salaries are at or above 819,5000
The middle 50% of salaries span approximately $417,500(819,500 - 402,000)

### quartile coefficent of dispersion

In [ ]:
quartile_coefficient = (q3 - q1) / (q3 + q1)

print(quartile_coefficient)
print(quartile_coefficient * 100, "%")

0.3417928776094965
34.17928776094965 %


This means the middle 50% of salaries have a relative dispersion of approximately 34.18%. There's a fair amount of variation between salaries in the middle half of the dataset - the salaries in the middle of the dataset are relatively spread out.

## Exercise 7: Scaling data
### min-max scaling

In [ ]:
min_max_scaled = [(salary - minimum_salary) / (maximum_salary - minimum_salary) for salary in salaries]
print(min_max_scaled)

[0.8472361809045226, 0.7608040201005025, 0.4221105527638191, 0.2592964824120603, 0.5125628140703518, 0.40603015075376886, 0.7869346733668342, 0.3035175879396985, 0.47839195979899496, 0.5849246231155779, 0.9115577889447236, 0.5065326633165829, 0.28241206030150756, 0.7587939698492462, 0.6201005025125628, 0.25125628140703515, 0.91356783919598, 0.9869346733668342, 0.8130653266331658, 0.9055276381909547, 0.31055276381909547, 0.7326633165829146, 0.9025125628140703, 0.6864321608040201, 0.4733668341708543, 0.10050251256281408, 0.43517587939698493, 0.6130653266331658, 0.9165829145728643, 0.9708542713567839, 0.47839195979899496, 0.8683417085427135, 0.26030150753768844, 0.8080402010050252, 0.5507537688442211, 0.01306532663316583, 0.7226130653266332, 0.4, 0.828140703517588, 0.6703517587939698, 0.0, 0.49547738693467336, 0.871356783919598, 0.2442211055276382, 0.3256281407035176, 0.8733668341708543, 0.19095477386934673, 0.5698492462311557, 0.23919597989949748, 0.9718592964824121, 0.8060301507537688, 

In [ ]:
print("Minimum scaled value:", min(min_max_scaled))
print("Maximum scaled value:", max(min_max_scaled))

Minimum scaled value: 0.0
Maximum scaled value: 1.0


Scaling turns each number into a value between 0 and 1; 0 being the lowest and 1 being the highest.

### standardizing

In [ ]:
standardized_salaries = [(salary - mean_manual)/ standard_deviation_manual for salary in salaries]
print(standardized_salaries)

[0.9717217942267801, 0.6482032533127501, -0.6195380058503674, -1.228956652688424, -0.28097209094033604, -0.6797275018343729, 0.7460111842867592, -1.0634355387324086, -0.40887476990634786, -0.010119359012310937, 1.2124797781628023, -0.3035431519343381, -1.142434252211416, 0.6406795663147493, 0.12154516345270126, -1.2590514006804265, 1.220003465160803, 1.4946180405878284, 0.8438191152607681, 1.1899087171688003, -1.037102634239406, 0.5428716353407403, 1.178623186671799, 0.36982683438672426, -0.4276839874013496, -1.8233279255304788, -0.5706340403633628, 0.09521225895969881, 1.2312889956578041, 1.434428544603823, -0.40887476990634786, 1.0507205077057873, -1.2251948091894236, 0.8250098977657664, -0.1380220379783228, -2.150608309943509, 0.5052532003507368, -0.7022985628283751, 0.9002467677457734, 0.3096373384027187, -2.199512275430514, -0.34492343042334195, 1.0620060382027885, -1.285384305173429, -0.9806749817544008, 1.0695297252007891, -1.4847620106204475, -0.06654701149731616, -1.3041935226

This tells us how far each salary is from the mean in standard-deviation units.
1 - one standard deviation above the mean; positive values are above the mean
-1 - one standard deviation below the mean; negative values are below the mean

In [ ]:
print("Mean:", statistics.mean(standardized_salaries))
print("Standard deviation:", statistics.stdev(standardized_salaries))

Mean: 3.86843335142828e-18
Standard deviation: 1.0


## Exercise 8: Calculating covariance and correlation
### covariance

In [ ]:
x_mean = statistics.mean(salaries)
y_mean = statistics.mean(min_max_scaled)
n = len(salaries)

covariance_manual = sum((x-x_mean) * (y-y_mean) for x, y in zip(salaries, min_max_scaled)) / (n-1)

print(covariance_manual)

71019.15019542155


In [ ]:
covariance_verified = statistics.covariance(salaries, min_max_scaled)
print(covariance_verified)

71019.15019542155


Covariance indicates whether 2 variables tend to move together:
Positive covariance: they increase together.
Negative covariance: they decrease together.
Near zero: no consistent linear movemnt

### Pearson correlation coefficient ($\rho$)

In [ ]:
correlation_manual = covariance_manual / (statistics.stdev(salaries) * statistics.stdev(min_max_scaled))
print(correlation_manual)

1.0


1.0 correlation means there is a perfect positive relationship.